builtin/providers/terraform/provider.go

```diff
@@ -1,21 +1,151 @@
+	DataSources map[string]providers.Schema
+}
+
+// NewProvider returns a new terraform provider
+func NewProvider() *Provider {
+	return &Provider{}
+}
+
+// GetSchema returns the complete schema for the provider.
+func (p *Provider) GetSchema() providers.GetSchemaResponse {
+	return providers.GetSchemaResponse{
+		Provider: providers.Schema{
+			Version: 1,
 		},
-		DataSourcesMap: map[string]*schema.Resource{
-			"terraform_remote_state": dataSourceRemoteState(),
+		DataSources: map[string]providers.Schema{
+			"terraform_remote_state": {
+				Block: &configschema.Block{
+					Attributes: map[string]*configschema.Attribute{
+						"backend": {
+							Type:     cty.String,
+							Required: true,
+						},
+						"config": {
+							Type:     cty.DynamicPseudoType,
+							Optional: true,
+						},
+						"defaults": {
+							Type:     cty.DynamicPseudoType,
+							Optional: true,
+						},
+						"outputs": {
+							Type:     cty.DynamicPseudoType,
+							Computed: true,
+						},
+						"workspace": {
+							Type:     cty.String,
+							Optional: true,
+						},
+					},
+				},
+			},
 		},
 	}
 }
+
+// ValidateProviderConfig is used to validate the configuration values.
+func (p *Provider) ValidateProviderConfig(providers.ValidateProviderConfigRequest) providers.ValidateProviderConfigResponse {
+	// At this moment there is nothing to configure for the terraform provider,
+	// so we will happily return without taking any action
+	var res providers.ValidateProviderConfigResponse
+	return res
+}
+
+// ValidateDataSourceConfig is used to validate the data source configuration values.
+func (p *Provider) ValidateDataSourceConfig(providers.ValidateDataSourceConfigRequest) providers.ValidateDataSourceConfigResponse {
+	// At this moment there is nothing to configure for the terraform provider,
+	// so we will happily return without taking any action
+	var res providers.ValidateDataSourceConfigResponse
+	return res
+}
+
+// Configure configures and initializes the provider.
+func (p *Provider) Configure(providers.ConfigureRequest) providers.ConfigureResponse {
+	// At this moment there is nothing to configure for the terraform provider,
+	// so we will happily return without taking any action
+	var res providers.ConfigureResponse
+	return res
+}
+
+// ReadDataSource returns the data source's current state.
+func (p *Provider) ReadDataSource(req providers.ReadDataSourceRequest) providers.ReadDataSourceResponse {
+	// call function
+	var res providers.ReadDataSourceResponse
+
+	// This should not happen
+	if req.TypeName != "terraform_remote_state" {
+		res.Diagnostics.Append(fmt.Errorf("Error: unsupported data source %s", req.TypeName))
+		return res
+	}
+
+	newState, diags := dataSourceRemoteStateRead(&req.Config)
+
+	res.State = newState
+	res.Diagnostics = diags
+
+	return res
+}
+
+// Stop is called when the provider should halt any in-flight actions.
+func (p *Provider) Stop() error {
+	panic("unimplemented")
```

Comment:
For this one, we should probably just make it a no-op (possibly logging "terraform provider can't Stop" or whatever), or else we'll probably get a panic if the user hits Ctrl+C while the data source is running.

Reply:
Yup, most providers don't actually do anything here, just returning nil for now is fine.